crates/jstzd/src/task/octez_client.rs

```diff
@@ -0,0 +1,199 @@
+use super::{endpoint::Endpoint, octez_node::DEFAULT_RPC_ENDPOINT};
+use anyhow::{bail, Result};
+use http::Uri;
+use std::{path::PathBuf, str::FromStr};
+use tempfile::{tempdir, TempDir};
+
+const DEFAULT_BINARY_PATH: &str = "octez-client";
+#[derive(Default)]
+pub struct OctezClientBuilder {
+    // if None, use the binary in $PATH
+    binary_path: Option<PathBuf>,
+    // if None, use temp directory
+    base_dir: Option<PathBuf>,
+    // if None, use localhost:8732 (the default endpoint for octez-node)
+    endpoint: Option<Endpoint>,
+    disable_unsafe_disclaimer: bool,
+}
+
+impl OctezClientBuilder {
+    pub fn new() -> Self {
+        OctezClientBuilder::default()
+    }
+    pub fn set_path(mut self, path: PathBuf) -> Self {
+        self.binary_path = Some(path);
+        self
+    }
+    pub fn set_base_dir(mut self, base_dir: PathBuf) -> Self {
+        self.base_dir = Some(base_dir);
+        self
+    }
+    pub fn set_endpoint(mut self, endpoint: Endpoint) -> Self {
+        self.endpoint = Some(endpoint);
+        self
+    }
+    pub fn set_disable_unsafe_disclaimer(
+        mut self,
+        disable_unsafe_disclaimer: bool,
+    ) -> Self {
+        self.disable_unsafe_disclaimer = disable_unsafe_disclaimer;
+        self
+    }
+    pub fn build(self) -> Result<OctezClient> {
+        self.validate_binary_path()?;
+        self.validate_base_dir()?;
+        let node_default_endpoint = format!("http://{}", DEFAULT_RPC_ENDPOINT);
+        Ok(OctezClient {
+            binary_path: self.binary_path.unwrap_or(DEFAULT_BINARY_PATH.into()),
+            base_dir: self
+                .base_dir
+                .map_or(Directory::TempDir(tempdir()?), Directory::Path),
+            endpoint: self
+                .endpoint
+                .unwrap_or(Endpoint::try_from(Uri::from_str(&node_default_endpoint)?)?),
+            disable_unsafe_disclaimer: self.disable_unsafe_disclaimer,
+        })
+    }
+    fn validate_binary_path(&self) -> Result<()> {
+        if let Some(binary_path) = &self.binary_path {
+            if !binary_path.exists() {
+                bail!("Binary path does not exist");
+            }
+            if !binary_path.is_file() {
+                bail!("Binary path is not a file");
+            }
+        }
+        Ok(())
+    }
+    fn validate_base_dir(&self) -> Result<()> {
+        if let Some(base_dir) = &self.base_dir {
+            if !base_dir.exists() {
+                bail!("Base directory does not exist");
+            }
+            if !base_dir.is_dir() {
+                bail!("Base directory is not a directory");
+            }
+        }
+        Ok(())
+    }
+}
+
+// TODO: remove later
+#[allow(dead_code)]
```

Comment:
This oculd be removed

Reply:
my ide complains about it so thought we needed it